# Report 
Group 5 - Product in consideration Diacetyl (mmol/gDW/h)\
Guilherme Lobo\
Karyna Lysenko\
Rodrigo Esperança

Biologia de Sistemas - Universidade do Minho\
dezembro de 2022

### Documentação (apagar depois)
https://cobrapy.readthedocs.io/en/latest/

In [5]:
import cobra
from cobra.io import read_sbml_model
from cobra.flux_analysis import flux_variability_analysis as fva

In [7]:
model = read_sbml_model("sth.xml")
#model

In [3]:
# print(len(model.reactions))
# print(len(model.metabolites))
# print(len(model.genes))

In [4]:
#To find out the Diacetyl metabolite id
for i in model.metabolites:
    if i.name == "Diacetyl_":
        print(i.id)

diact_c
diact_e


In [5]:
diact_c=model.metabolites.get_by_id("diact_c")

In [7]:
diact_e=model.metabolites.get_by_id("diact_e")

In [6]:
DIACTt = model.reactions.get_by_id("DIACTt")
DIACTt

<Reaction DIACTt at 0x249ee6e6e20>

In [8]:
casp = model.reactions.get_by_id("EX_caspep_e")
casp

<Reaction EX_caspep_e at 0x2258d68cb80>

In [8]:
EX_diact_e = model.reactions.get_by_id("EX_diact_e")
EX_diact_e

<Reaction EX_diact_e at 0x249ee6f1ee0>

## Chemical defined medium (CDM)
### Function objetive: maximization of biomass production.

##### 1 a) What is the wild-type production of the compound?


In order to have a general idea of the flux of biomass_STR and of the exchange reactions we applied the summary() function and noted the biomass production is 1.042 1/h.

In [7]:
model.summary()
# model.slim_optimize()

In [13]:
model.optimize()["EX_diact_e"] # optimize só dá fluxos dos exchanges

0.0

In [8]:
#model.metabolites.diact_e.summary()
model.metabolites.diact_c.summary()

Response:
The wild-type production of the Diacetyl is 0.0 1/h.

#### 1 b) Access the robustness of the presented solution using the Flux Variability Analysis approach.

In [14]:
#Result of FBA = 1.042 (units ?)
model_fva=fva(model,fraction_of_optimum=1)

In [15]:
model_fva.loc["EX_diact_e"] #maximo mto baixo, logo é como 0

minimum    0.000000e+00
maximum    8.254372e-13
Name: EX_diact_e, dtype: float64

Conclusion: As the minimum and maximum are very close to each other (maximum is almost 0.0), we can conclude that the solution is robust.

#### 1 c) What are the maximum compound production capabilities, guaranteeing a minimum growth rate of 20% of the wild type?

In [18]:
model_1c = read_sbml_model("sth.xml")

In [19]:
model_1c.summary(fva=0.20)

In [35]:
model_1c.optimize()["EX_diact_e"]

0.0

Conclusion:

#### 1 d) Evaluate if single gene deletions enhance the production of the compound. Rank the mutants obtained according to the compound production capacity and growth performance.

In [3]:
from cobra.flux_analysis import (single_gene_deletion, double_gene_deletion)

In [21]:
#original
model_knockout = read_sbml_model("sth.xml")
for i in model_knockout.genes:
    with model_knockout:
        i.knock_out()
        tmp=model_knockout.optimize()
        #print(tmp.fluxes, tmp.status) #producao do composto + taxa de crescimento
        print(model_knockout.optimize()["EX_diact_e"])

0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0


C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0


C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0


C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)



0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)



0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning


0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)



0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0


C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0


C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


Conclusion: The deletion of each gene doesn't affect the Diacetil production, the flux remains 0.0 mmol/gDW/h.

In [13]:
#to check the balance
#EX_diact_e.check_mass_balance()

{'C': -4.0, 'H': -6.0, 'O': -2.0}

In [23]:
test = read_sbml_model("sth.xml")
print(single_gene_deletion(test,test.genes).info)

## Milk medium (MM)
### Function objetive: minimization of casein consumption while constraining the growth rate to 0.64 h-1

In [20]:
model_milk.optimize()["EX_caspep_e"] #

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


0.64

In [ ]:
#To find out the metabolite id
for i in model.metabolites:
    print(i)

In [9]:
model_milk = read_sbml_model("sth.xml")
model_milk.objective = 'EX_caspep_e' #colocar igual ao id anterior
#model_milk.summary()


In [44]:
#model_milk.medium

In [14]:
#originals
# print(model_milk.metabolites.lac_L_c.summary()) #lactato
# model_milk.reactions.EX_lac_L_e
# print(model_milk.metabolites.lcts_e.summary()) #lactose
# model_milk.reactions.EX_lcts_e
# print(model_milk.metabolites.gal_e.summary()) #galactose
# model_milk.reactions.EX_gal_e
# print(model_milk.metabolites.glc_D_e.summary()) #glusoce
# model_milk.reactions.EX_glc_D_e
# print(model_milk.metabolites.caspep_e.summary()) #cas
# model_milk.reactions.EX_caspep_e
# model_milk.reactions.EX_mpept_e #mpep

# to discover the id's of the aminoacids
# for i in model.metabolites:
#     print(i.name, i.id)

In [12]:
#change the conditions of the solution
model_milk.reactions.EX_lac_L_e.lower_bound=30
model_milk.reactions.EX_lac_L_e.upper_bound=40.8

model_milk.reactions.EX_lcts_e.lower_bound= (-25.1)
model_milk.reactions.EX_lcts_e.upper_bound= (-22.6)

model_milk.reactions.EX_gal_e.lower_bound=0
model_milk.reactions.EX_gal_e.upper_bound=1000

model_milk.reactions.EX_glc_D_e.lower_bound=2
model_milk.reactions.EX_glc_D_e.upper_bound=2.5

model_milk.reactions.EX_caspep_e.lower_bound=-1000
model_milk.reactions.EX_caspep_e.upper_bound=1000

model_milk.reactions.EX_mpept_e.lower_bound=-1000
model_milk.reactions.EX_mpept_e.upper_bound=1000

In [13]:
model_milk.reactions.EX_arg_L_e.lower_bound=0
model_milk.reactions.EX_arg_L_e.upper_bound=0

model_milk.reactions.EX_asp_L_e.lower_bound=0
model_milk.reactions.EX_asp_L_e.upper_bound=0

model_milk.reactions.EX_cys_L_e.lower_bound=0
model_milk.reactions.EX_cys_L_e.upper_bound=0

model_milk.reactions.EX_his_L_e.lower_bound=0
model_milk.reactions.EX_his_L_e.upper_bound=0

model_milk.reactions.EX_lys_L_e.lower_bound=0
model_milk.reactions.EX_lys_L_e.upper_bound=0

model_milk.reactions.EX_gln_L_e.lower_bound=0
model_milk.reactions.EX_gln_L_e.upper_bound=0

model_milk.reactions.EX_ser_L_e.lower_bound=0
model_milk.reactions.EX_ser_L_e.upper_bound=0

# model_milk.reactions.Ex_ser_D_e.lower_bound=0
# model_milk.reactions.EX_ser_D_e.upper_bound=0

model_milk.reactions.EX_thr_L_e.lower_bound=0
model_milk.reactions.EX_thr_L_e.upper_bound=0

model_milk.reactions.EX_glu_L_e.lower_bound=0
model_milk.reactions.EX_glu_L_e.upper_bound=0

model_milk.reactions.EX_glyc_e.lower_bound=0
model_milk.reactions.EX_glyc_e.upper_bound=0

model_milk.reactions.EX_pro_L_e.lower_bound=0
model_milk.reactions.EX_pro_L_e.upper_bound=0

model_milk.reactions.EX_ala_L_e.lower_bound=0
model_milk.reactions.EX_ala_L_e.upper_bound=0

model_milk.reactions.EX_ala_D_e.lower_bound=0
model_milk.reactions.EX_ala_D_e.upper_bound=0

model_milk.reactions.EX_val_L_e.lower_bound=0
model_milk.reactions.EX_val_L_e.upper_bound=0

model_milk.reactions.EX_ile_L_e.lower_bound=0
model_milk.reactions.EX_ile_L_e.upper_bound=0

model_milk.reactions.EX_leu_L_e.lower_bound=0
model_milk.reactions.EX_leu_L_e.upper_bound=0

model_milk.reactions.EX_met_L_e.lower_bound=0
model_milk.reactions.EX_met_L_e.upper_bound=0

# model_milk.reactions.EX_met_D_e.lower_bound=0
# model_milk.reactions.EX_met_D_e.upper_bound=0

model_milk.reactions.EX_phe_L_e.lower_bound=0
model_milk.reactions.EX_phe_L_e.upper_bound=0

model_milk.reactions.EX_tyr_L_e.lower_bound=0
model_milk.reactions.EX_tyr_L_e.upper_bound=0

model_milk.reactions.EX_trp_L_e.lower_bound=0
model_milk.reactions.EX_trp_L_e.upper_bound=0

In [14]:
model_milk.reactions.EX_caspep_e.lower_bound=0.64
model_milk.reactions.EX_caspep_e.upper_bound=0.64
# the lower_bound has to be 0. If we set as 0.64, the model is infeasable

In [17]:
casp_m = model_milk.reactions.get_by_id("EX_caspep_e")
casp_m

<Reaction EX_caspep_e at 0x2258d8316d0>

#### 1 a) What is the wild-type production of the compound?

In [15]:
model_milk.summary()

Infeasible: None (infeasible).

In [25]:
model_milk.optimize()["EX_diact_e"]

1.811280632411065

Conclusion: the flux of Diacetil is 1.811 1/h.

#### 1 b) Access the robustness of the presented solution using the Flux Variability Analysis approach.

In [26]:
model_milk_fva=fva(model_milk,fraction_of_optimum=1)
model_milk_fva.loc["EX_diact_e"] 

minimum    0.00
maximum    9.17
Name: EX_diact_e, dtype: float64

The result of the last FBA is not robust as it's possible to acheive a maximum flux for Diacetil of 9.17 1/h.

#### 1 c) What are the maximum compound production capabilities, guaranteeing a minimum growth rate of 20% of the wild type?

In [ ]:
# model_milk.reactions.biomass_STR.lower_bound=???
# model_milk.reactions.biomass_STR.upper_bound=???
# model_milk.summary()

#### 1 d) Evaluate if single gene deletions enhance the production of the compound. Rank the mutants obtained according to the compound production capacity and growth performance.